### ISSUE TITLE
Adding compression

### ISSUE THREAD

**Issue by rimmeraj** [RuedigerMoeller/fast-serialization#132]
I have a situation where I am serializing a a large Map with nested objects. What is happening is the client is running out of memory. I see in FSTOutputStream you are serializing the object graph to a byte[] and then writing to a passed in stream. Is there any way we could add a threshold so if the byte array crossed that threshold it was flushed to the output stream and reset? That way we could be using some sort of Compression Stream to pack the chunks.

**Comment by RuedigerMoeller** [comment]
The major consumer of memory is your huge map, as the unserialized map on-the-heap will consume like 10 times the memory compared to its serialized form in the serialization buffer. So adding compression won't really help.
Try to reduce footprint of your map e.g. by using more memory efficient data structures. E.g. open addressing hash maps instead of java.lang.HMap.

It does not make sense to add a major slowdown like compression magically (an even having an additional check with each byte written is a no no). 

What could be done is a FSTCompressedDefaultConfiguration or so.

**Comment by rimmeraj** [comment]
I do not believe in magic, that is not where I was going. What I would be looking for is for the FSTOutputStream buffer to simply flush to the given outputstream when it got to a certain threshold and then reset it's buffer. That way you would have a fixed size buffer and the output stream could either be pushing packets over a wire, or compressing them. Ideally after serialization you could ask FSTOutputStream if it hit it's threshold. If it did , flush the rest of the buffer to the outputstream and that is where your result is , or just grab the byte buffer and do not flush.  Maybe an option to use a FSTThresholdOutputStream ( class extending FSTOutputStream with this functionality) ?

**Comment by RuedigerMoeller** [comment]
In theory you are right regarding premature stream flushing ..
however flushing cannot be implemented easily as stream position of an object is used to mark and resolve links inside the object graph, so all code dealing with that needs to take into account intermediate flushing's effect on streamposition and object ids. This is a todo since 2.0 (1.X got that right). As there are way more edge cases in supporting stock JDK serialization methods (read/writeObject, writeFields, readResolve, stock stream implementations) than even an experienced developer would expect, it's really easy to break things, so I shyed away from implementing this ..

**Comment by abyrd** [comment]
I just wanted to let you know that another user (me) is interested in this feature. I also have a situation where a fixed-size output buffer would be ideal, and intermediate flushing seems like the relevant approach. The object we are serializing contains references to a bunch of primitive arrays. Java's own representation of these arrays is not much larger than the serialized version. Serializing such an object nearly doubles memory consumption of our application as the FST buffer expands to contain the serialized representation of the entire object tree.

In the case I describe, we are using serialization as a way to persist large data structures to disk, so another alternative is to use memory mapped files. I successfully created and applied a MemMapStreamCoderFactory, but I see no convenient way to resize the mapped file and associated buffer to match the size of the object to be serialized.
